python/pyspark/mllib/clustering.py

```diff
@@ -88,6 +92,162 @@ def train(cls, rdd, k, maxIterations=100, runs=1, initializationMode="k-means||"
+
+        :param x: a vector
+        :return: the closest cluster index
+        """
+        ser = PickleSerializer()
+        bytes = bytearray(ser.dumps(_convert_to_vector(x)))
+        vec = self._sc._jvm.SerDe.loads(bytes)
+        result = self._java_model.predict(vec)
+        return PickleSerializer().loads(str(self._sc._jvm.SerDe.dumps(result)))
+
+    def __predict_by_rdd(self, x):
+        """Predict the closest cluster index with a RDD
+        :param x: a RDD of vector
+        :return: a RDD of int
+        """
+        ser = PickleSerializer()
+        cached = x.map(_convert_to_vector)._reserialize(AutoBatchedSerializer(ser)).cache()
+        rdd = _to_java_object_rdd(cached)
+        jrdd = self._java_model.predict(rdd)
+        jpyrdd = self._sc._jvm.SerDe.javaToPython(jrdd)
+        return RDD(jpyrdd, self._sc, AutoBatchedSerializer(PickleSerializer()))
+
+    def cut(self, height):
+        """Cut nodes and leaves in a cluster tree by a dendrogram height.
+        :param height: a threshold to cut a cluster tree
+        """
+        ser = PickleSerializer()
+        model = self._java_model.cut(height)
+        bytes = self._sc._jvm.SerDe.dumps(model.getCenters())
+        centers = ser.loads(str(bytes))
+        return HierarchicalClusteringModel(self._sc, model, [c.toArray() for c in centers])
+
+    def sum_of_variance(self):
+        """Gets the sum of variance of all clusters.
+        :return: sum of variance of all clusters
+        """
+        ser = PickleSerializer()
+        model = self._java_model
+        bytes = self._sc._jvm.SerDe.dumps(model.getSumOfVariance())
+        variance = ser.loads(str(bytes))
+        return variance
+
+    def to_merge_list(self):
+        """Extract an array for dendrogram
+
+        the array is fit for SciPy's dendrogram
+        :return: an array which is fit for scipy's dendrogram
+        """
+        ser = PickleSerializer()
+        model = self._java_model
+        bytes = self._sc._jvm.SerDe.dumps(model.toMergeList())
+        centers = ser.loads(str(bytes))
+        return array([c.toArray() for c in centers])
+
+
+class HierarchicalClustering(object):
+
+    @classmethod
+    def train(cls, rdd, k,
+              subIterations=20, numRetries=10, epsilon=1.0e-4, randomSeed=1, randomRange=0.1):
+        """Train a hierarchical clustering model."""
+        sc = rdd.context
+        ser = PickleSerializer()
+        # cache serialized data to avoid objects over head in JVM
+        cached = rdd.map(_convert_to_vector)._reserialize(AutoBatchedSerializer(ser)).cache()
+        model = sc._jvm.PythonMLLibAPI().trainHierarchicalClusteringModel(
+            _to_java_object_rdd(cached), k,
+            subIterations, numRetries, epsilon, randomSeed, randomRange)
+        bytes = sc._jvm.SerDe.dumps(model.getCenters())
+        centers = ser.loads(str(bytes))
+        # TODO: because centers are SparseVector, wel will change them numpy.array.
```

Comment:
Did you mean convert from `DenseVector`? That's currently the output. I agree it would be good to change to `numpy.array`, especially because after calling `cut` the output is an array of `numpy.array`s. The two behaviors should be consistent, as calling `cut` should not change the type.